data-plane/dispatcher-loom/src/main/java/dev/knative/eventing/kafka/broker/dispatcherloom/LoomKafkaConsumer.java

```diff
@@ -15,6 +15,177 @@
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import dev.knative.eventing.kafka.broker.core.ReactiveKafkaConsumer;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+
+public class LoomKafkaConsumer<K, V> implements ReactiveKafkaConsumer<K, V> {
+
+    private static final Logger logger = LoggerFactory.getLogger(LoomKafkaConsumer.class);
+
+    private final Consumer<K, V> consumer;
+    private final BlockingQueue<Runnable> taskQueue;
+    private final AtomicBoolean isClosed;
+    private final Thread taskRunnerThread;
+    private Handler<Throwable> exceptionHandler;
+
+    public LoomKafkaConsumer(Vertx vertx, Consumer<K, V> consumer) {
+        this.consumer = consumer;
+        this.taskQueue = new LinkedBlockingQueue<>();
+        this.isClosed = new AtomicBoolean(false);
+
+        taskRunnerThread = Thread.ofVirtual().start(this::processTaskQueue);
+    }
+
+    private void addTask(Runnable task, Promise<?> promise) {
+        if(isClosed.get()) {
+            promise.fail("Consumer is closed");
+        }
+        taskQueue.add(task);
+    }
+
+    private void processTaskQueue() {
+        while(!isClosed.get() || !taskQueue.isEmpty()){
+            try {
+                taskQueue.take().run();
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting for task", e);
+                exceptionHandler.handle(e);
+            }
+        }
+    }
+
+    @Override
+    public Future<Map<TopicPartition, OffsetAndMetadata>> commit(Map<TopicPartition, OffsetAndMetadata> offset) {
+        Promise<Map<TopicPartition, OffsetAndMetadata>> promise = Promise.promise();
+        addTask(() -> {
+                consumer.commitAsync(offset, (offsetMap, exception) -> {
+                    if(exception != null) {
+                        promise.fail(exception);
+                    } else {
+                        promise.complete(offsetMap);
+                    }
+                });
+        }, promise);
+        return promise.future();
+    }
+
+    @Override
+    public Future<Void> close() {
+        Promise<Void> promise = Promise.promise();
+        isClosed.set(true);
+        Thread.ofVirtual().start(() -> { // I am not sure we can call close from diffrent thread
+            try {
+                while(!taskQueue.isEmpty()) {
+                    Thread.sleep(100);
+                }
+                taskRunnerThread.interrupt();
+                taskRunnerThread.join();
+                consumer.close();
+                promise.complete();
+            } catch (Exception e) {
+                promise.fail(e);
+            }
+        });
+        return promise.future();
+    }
```

Comment:
Definitely need suggestions here.

Reply:
Yes, we cannot do that, we need to queue the "close task" and then on a separate thread we can 
```
        // add close task to queue ...

        // wait for tasks to complete
        Thread.ofVirtual().start(() -> {
            try {
                while(!taskQueue.isEmpty()) {
                    Thread.sleep(2000);
                }
                taskRunnerThread.interrupt();
                taskRunnerThread.join();
                promise.complete();
            } catch (Exception e) {
                promise.fail(e);
            }
        });```